clippy_utils/src/ast_utils.rs

```diff
@@ -571,3 +572,39 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
         _ => false,
     }
 }
+
+/// Extract args from an assert-like macro.
+/// Currently working with:
+/// - `assert!`, `assert_eq!` and `assert_ne!`
+/// - `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!`
+/// For example:
+/// `assert!(expr)` will return Some([expr])
+/// `debug_assert_eq!(a, b)` will return Some([a, b])
+pub fn extract_assert_macro_args(mut expr: &Expr) -> Option<Vec<&Expr>> {
+    if_chain! {
+        if let ExprKind::If(_, ref block, _) = expr.kind;
+        if let StmtKind::Semi(ref e) = block.stmts.get(0)?.kind;
+        then {
+            expr = e;
+        }
+    }
+    if_chain! {
+        if let ExprKind::Block(ref block, _) = expr.kind;
+        if let StmtKind::Expr(ref expr) = block.stmts.get(0)?.kind;
+        if let ExprKind::Match(ref match_expr, _) = expr.kind;
+        if let ExprKind::Tup(ref tup) = match_expr.kind;
+        then {
+            if let [a, b, ..] = tup.as_slice() {
+                if let (&ExprKind::AddrOf(_, _, ref a), &ExprKind::AddrOf(_, _, ref b)) = (&a.kind, &b.kind) {
+                    return Some(vec![&*a, &*b]);
+                }
+            }
+            if let [a, ..] = tup.as_slice() {
```

Comment:
Isn't the `, ..` redundant? It looks like this would be better written as a single `match &**tup { [a] => .., [a, b, ..] => .., }`.

Reply:
It was in the case not both arguments were `AddrOf`, so we handle both in case there is only one argument and when there are two where the second is invalid for whatever reason.

Reply:
It could be problematic to return one arg when there is actually more. But if we're just working on hypotheticals, I would rather err on the side of simplicity.

Reply:
Then I'll revert and explicitly not support `assert!`.